lib/msal-common/src/server/ServerRequestParameters.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { ICrypto } from "../crypto/ICrypto";
+import { Constants } from "../utils/Constants";
+import pkg from "../../package.json";
+import { ScopeSet } from "../auth/ScopeSet";
+
+export abstract class ServerRequestParameters {
+    // Crypto functions
+    protected cryptoObj: ICrypto;
+
+    // Params
+    clientId: string;
+    scopes: ScopeSet;
+    redirectUri: string;
+
+    // Validity checks
+    state: string;
+
+    // Telemetry Info
+    xClientVer: string;
+    xClientSku: string;
+    correlationId: string;
+
+    constructor(clientId: string, redirectUri: string, cryptoImpl: ICrypto) {
+        this.clientId = clientId;
+        this.cryptoObj = cryptoImpl;
+        this.redirectUri = redirectUri;
+
+        // Telemetry Info
+        this.xClientSku = Constants.LIBRARY_NAME;
+        this.xClientVer = pkg.version;
+    }
+
+    protected abstract createParamString(): Promise<Array<string>>;
+    
+    /**
+     * Replace client id with the default scopes used for token acquisition.
+     */
+    protected replaceDefaultScopes() {
```

Comment:
What's the use case for this? Shouldn't our default scopes... just work to begin with? Or does a second call require different scopes?

Reply:
if the client id is present in the list of scopes, we remove it and add the openid and profile scopes to retrieve the id token. If it is only an access token (so no clientid in list of scopes), we don't need to add openid and profile. We ALWAYS need the offline_access scope in order to get the refresh token for SPAs.